crates/sui-framework/packages/move-stdlib/sources/vector.move

```diff
@@ -361,4 +361,90 @@ module std::vector {
         zip_do_ref!($v1, $v2, |el1, el2| r.push_back($f(el1, el2)));
         r
     }
+
+    /// Iterate through the vector and find the indices of elements that satisfy the predicate `p`.
+    /// Returns a new vector containing the indices of elements that satisfy the predicate in the
+    /// order they appear in the original vector.
+    public macro fun find_indices<$T>($v: &vector<$T>, $p: |&$T| -> bool): vector<u64> {
+        let v = $v;
+        let mut r = vector[];
+        v.length().do!(|i| if ($p(&v[i])) r.push_back(i));
+        r
+    }
+
+    /// Take the first `n` elements of the vector `v`. Aborts if `n` is greater than the length of `v`.
+    /// Modifies the original vector.
+    public macro fun take<$T>($v: &mut vector<$T>, $n: u64): vector<$T> {
+        let n = $n;
+        let v = $v;
+        assert!(n <= v.length());
+        v.reverse();
+        let mut r = vector[];
+        n.do!(|_| r.push_back(v.pop_back()));
+        v.reverse();
+        r
+    }
+
+    /// Takes the first `n` elements of the vector `v` that satisfy the predicate `p`.
+    /// Modifies the original vector.
+    public macro fun take_while<$T>($v: &mut vector<$T>, $p: |&$T| -> bool): vector<$T> {
+        let v = $v;
+        let mut r = vector[];
+        v.reverse();
+        while (v.length() != 0 && $p(&v[v.length() - 1])) {
+            r.push_back(v.pop_back());
+        };
+        v.reverse();
+        r
+    }
+
+    /// Take the first `n` elements of the vector `v` and drop the rest. Aborts if `n` is greater
+    /// than the length of `v`. Destroys the original vector after taking the elements.
+    public macro fun take_and_drop<$T: drop>($v: vector<$T>, $n: u64): vector<$T> {
```

Comment:
A thought though as time has passed on these names:
None of our other compounded functions (like `zip_map` or something) have `_and_` in the middle. And `filter` isn't `filter_drop`. So maybe they should be
```
fun take_mut<T>
fun take<T: drop>
macro fun take_while<$T: drop>
macro fun take_mut_while<$T>
```
And similarly for `skip*`

Reply:
Love this, let's totally do.